File: src/main/java/com/khlud/ciprian/flatcollection/codegen/CodeGenerator.java
package com.khlud.ciprian.flatcollection.codegen;

import spark.ModelAndView;
import spark.template.mustache.MustacheTemplateEngine;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Ciprian on 1/23/2016.
 */
public class CodeGenerator {
    FlatClassDescription _description;
    MustacheTemplateEngine _engineArrayList = new MustacheTemplateEngine();

    public CodeGenerator(FlatClassDescription description) {
        _description = description;

    }

    static final String intToString(int value) {
        Integer val = value;
        return val.toString();
    }

    public String generateArrayList() {
        Map<String, String> valuesView = new HashMap<>();

        valuesView.put("typeName", _description.getSimpleName());
        valuesView.put("valueType", _description.getFieldTypeName());
        valuesView.put("packageName", "FlatCollections");
        valuesView.put("countFields", intToString(_description.getFieldNames().size()));

        ModelAndView modelAndView = new ModelAndView(valuesView, "ArrayListOfValue.mustache");
        String iteratorCode = _engineArrayList.render(modelAndView);
        StringBuilder sb = new StringBuilder();
        sb.append(iteratorCode);

        return sb.toString();
    }

    public String generateIterator() {
        String fieldSection = generate();
        Map<String, String> valuesView = new HashMap<>();

        valuesView.put("typeName", _description.getSimpleName());
        valuesView.put("valueType", _description.getFieldTypeName());
        valuesView.put("packageName", "FlatCollections");
        valuesView.put("countFields", intToString(_description.getFieldNames().size()));

        ModelAndView modelAndView = new ModelAndView(valuesView, "flatCursor.mustache");
        String iteratorCode = _engineArrayList.render(modelAndView);
        StringBuilder sb = new StringBuilder();
        sb.append(iteratorCode);

        sb.append(fieldSection);
        sb.append("\n}\n");
        return sb.toString();
    }

    public String generate() {
        StringBuilder sb = new StringBuilder();
        int[] index = {0};
        Map<String, String> valuesView = new HashMap<>();
        valuesView.put("typeField", _description.getFieldTypeName());
        _description.getFieldNames().forEach(field -> {
            valuesView.put("fieldName", field);
            valuesView.put("index", intToString(index[0]));
            ModelAndView modelAndView = new ModelAndView(valuesView, "fieldGetterSetter.mustache");
            String fieldText = _engineArrayList.render(modelAndView);
            sb.append(fieldText);

            index[0]++;

        });

        return sb.toString();
    }
}

File: src/main/java/com/khlud/ciprian/flatcollection/codegen/CodeGenerator.kt
package com.khlud.ciprian.flatcollection.codegen

import spark.ModelAndView
import spark.template.mustache.MustacheTemplateEngine
import java.util.*

/**
 * Created by Ciprian on 1/23/2016.
 */
class CodeGenerator(val _description: FlatClassDescription) {
    internal var _engineArrayList = MustacheTemplateEngine()

    private fun intToString(value: Int): String {
        return value.toString()
    }

    fun generateArrayList(): String {
        val valuesView: MutableMap<String, String> = HashMap()

        valuesView["typeName"] = _description.simpleName
        valuesView["valueType"] = _description.fieldTypeName
        valuesView["packageName"] = "FlatCollections"
        valuesView["countFields"] = intToString(_description.fieldNames.size)

        val modelAndView = ModelAndView(valuesView, "ArrayListOfValue.mustache")
        val iteratorCode = _engineArrayList.render(modelAndView)
        val sb = StringBuilder()
        sb.append(iteratorCode)

        return sb.toString()
    }

    fun generateIterator(): String {
        val fieldSection = generate()
        val valuesView = HashMap<String, String>()

        valuesView["typeName"] = _description.simpleName
        valuesView["valueType"] = _description.fieldTypeName
        valuesView["packageName"] = "FlatCollections"
        valuesView["countFields"] = intToString(_description.fieldNames.size)

        val modelAndView = ModelAndView(valuesView, "flatCursor.mustache")
        val iteratorCode = _engineArrayList.render(modelAndView)

        return iteratorCode +
                fieldSection +
                "\n}\n"
    }

    fun generate(): String {
        val sb = StringBuilder()
        val index = intArrayOf(0)
        val valuesView = HashMap<String, String>()
        valuesView["typeField"] = _description.fieldTypeName
        _description.fieldNames.forEach { field ->
            valuesView["fieldName"] = field
            valuesView["index"] = intToString(index[0])
            val modelAndView = ModelAndView(valuesView, "fieldGetterSetter.mustache")
            val fieldText = _engineArrayList.render(modelAndView)
            sb.append(fieldText)

            index[0]++

        }

        return sb.toString()
    }
}